xmtp_mls/src/groups/device_sync/message_sync.rs

```diff
@@ -0,0 +1,291 @@
+use super::*;
+use crate::storage::group_message::MsgQueryArgs;
+use crate::XmtpApi;
+use crate::{storage::group::StoredGroup, Client};
+use xmtp_id::scw_verifier::SmartContractSignatureVerifier;
+
+impl<ApiClient, V> Client<ApiClient, V>
+where
+    ApiClient: XmtpApi + Clone,
+    V: SmartContractSignatureVerifier + Clone,
+{
+    // returns (request_id, pin_code)
+    pub async fn send_history_sync_request(&self) -> Result<(String, String), DeviceSyncError> {
+        let request = DeviceSyncRequest::new(DeviceSyncKind::MessageHistory);
+        self.send_sync_request(request).await
+    }
+
+    pub async fn reply_to_history_sync_request(
+        &self,
+    ) -> Result<DeviceSyncReplyProto, DeviceSyncError> {
+        let (_msg, request) = self
+            .pending_sync_request(DeviceSyncKind::MessageHistory)
+            .await?;
+
+        let groups = self.syncable_groups()?;
+        let messages = self.syncable_messages()?;
+
+        let reply = self
+            .create_sync_reply(
+                &request.request_id,
+                &[groups, messages],
+                DeviceSyncKind::MessageHistory,
+            )
+            .await?;
+        self.send_sync_reply(reply.clone()).await?;
+
+        Ok(reply)
+    }
+
+    pub async fn process_history_sync_reply(&self) -> Result<(), DeviceSyncError> {
+        self.process_sync_reply(DeviceSyncKind::MessageHistory)
+            .await
+    }
+
+    fn syncable_groups(&self) -> Result<Vec<Syncable>, DeviceSyncError> {
+        let conn = self.store().conn()?;
+        let groups = conn
+            .find_groups(None, None, None, None, Some(ConversationType::Group))?
+            .into_iter()
+            .map(Syncable::Group)
+            .collect();
+        Ok(groups)
+    }
+
+    fn syncable_messages(&self) -> Result<Vec<Syncable>, DeviceSyncError> {
+        let conn = self.store().conn()?;
+        let groups = conn.find_groups(None, None, None, None, Some(ConversationType::Group))?;
+
+        let mut all_messages = vec![];
+        for StoredGroup { id, .. } in groups.into_iter() {
+            let messages = conn.get_group_messages(&id, &MsgQueryArgs::default())?;
+            for msg in messages {
+                all_messages.push(Syncable::GroupMessage(msg));
+            }
+        }
+
+        Ok(all_messages)
+    }
+}
+
+#[cfg(all(not(target_arch = "wasm32"), test))]
```

Comment:
sucks that we don't have any history sync tests for wasm, I created a ticket to eventually create some integration tests, when the message-history/consent sync server is ever added to docker-compose: https://github.com/xmtp/libxmtp/issues/1201

I skimmed some http-mock libraries, but none have wasm support b/c of the nature of http, so the docker route is probably the only way to go